dali/sasha/saarch.cpp

```diff
@@ -83,297 +83,449 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
-                                pt.getProp("@jobName",val);
-                            else if (stricmp(tok,"state")==0)
-                                pt.getProp("@state",val);
-                            else
-                                found = false;
-                            if (found)
-                                name.append(',').append(val);
-                            tok = strtok_r(NULL, "|,",&saveptr);
-                        }
-                        free(parse);
+
+                    if (sortInc)
+                    {//The output should contain WUs from new to old. So, we need to reverse the order before output.
+                        Owned<CWUData> wu = new CWUData();
+                        wu->wuTree.setown(wuTree.getClear());
+                        if (outputModifiedTime)
+                            fileIterator->getModifiedTime(wu->modifiedTime);
+                        wus.append(*wu.getClear());
                     }
-                    cmd->addId(name.str());
-                    if (cmd->getAction()==SCA_GET) {
-                        StringBuffer xml;
-                        toXML(&pt,xml);
-                        if (!cmd->addResult(xml.str()))
+                    else
+                    {
+                        CDateTime dt;
+                        if (outputModifiedTime)
+                            fileIterator->getModifiedTime(dt);
+
+                        if (!addOutput(outputFields, outputModifiedTime, hasWUSOutput, wuTree, dt, WUSbuf))
+                        {
+                            overflowed = true;
                             break;
+                        }
                     }
+                    wuCount++;
+                    if (wuCount == maxNumWUs)
+                        break;
+                }
+                if (overflowed || (wuCount == maxNumWUs))
+                    break;
+            }
+            if (sortInc)
+            {
+                ForEachItemInRev(i, wus)
+                {
+                    CWUData &wuData = wus.item(i);
+                    if (!addOutput(outputFields, outputModifiedTime, hasWUSOutput, wuData.wuTree, wuData.modifiedTime, WUSbuf))  //Log an error?
+                        break;
                 }
-                if (index>start+num)
+            }
+
+            if (hasWUSOutput)
+                cmd->setWUSresult(WUSbuf);
+        }
+        void getOnlineWUs()
+        {
+            if (cmd->getAction()==SCA_WORKUNIT_SERVICES_GET)
+                throw MakeStringException(-1,"SCA_WORKUNIT_SERVICES_GET not implemented for online workunits!");
+
+            Owned<IRemoteConnection> conn = getSDSConnection(mask.str());
+            if (!conn)
+                return;
+    
+            IArrayOf<IPropertyTree> wus;
+            unsigned wuCount = 0;
+            StringBuffer wuid;
+            Owned<IPropertyTreeIterator> iter = conn->queryRoot()->getElements(isWild ? "*" : NULL);
```

Comment:
```
jakesmith 5 days ago  Member
Not new, but shouldn't this getElements being passing the mask if isWild instead of "*"?
i.e. get the property tree impl. to pre-filter - rather than get all and filter in parseAndCheckWUID with WildMatch

If so, could be a separate JIRA to address this.
  @wangkx
wangkx 5 days ago  Member
We do not have to check the mask in both the getElements and the parseAndCheckWUID because the mask has been used in the getSDSConnection (line 469).
```

>because the mask has been used in the getSDSConnection (line 469).

There's something not right here..
It can't connect to the mask, e.g. connecting to /WorkUnits/W201701* will fail.

If a wild card is involved, it must connect to the root and filter.
And looking the current code (master) that's what it does. It only uses the mask if connect is it is *not* wild.

The same was supposed to be true here and it will fail with an getOnlineWUs call with a wild card as it is, afaics.
It should be : 
```
    Owned<IRemoteConnection> conn = getSDSConnection(isWild ? nullptr : mask.str());
```

Then, filtering _is_ needed .. and in the old code was done by WildMatch (line 306 in master), but could be done by getElements more directly.

Reply:
Will fix.